docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+
+  Cursors are time-ordered so that given event items A and B, if A was
+  published before B, then cursor(A) < cursor(B) in lexicographic order.
+
+  A minimum viable cursor implementation is a tuple consisting of a timestamp
+  and a sequence number (e.g., `16CCC798FB5F4670-0123`). However, it may also
+  be useful to append basic type information to a cursor, to allow efficient
+  filtering (e.g., `16CCC87E91869050-0091:BeginBlock`).
+
+  The initial implementation will use the minimum viable format.
+
+### Discussion
+
+The node maintains an **event log**, a shared ordered record of the events
+published to its event bus within an operator-configurable time window.  The
+initial implementation will store the event log in-memory, and the operator
+will be given two configuration settings.  Note, these names are provisional:
+
+- `[event-subscription] time-window`: A duration before present during which the
+  node will retain event items published. Setting this value to zero disables
+  event subscription.
+
+- `[event-subscription] max-items`: A maximum number of event items that the
+  node will retain within the time window. If the number of items exceeds this
+  value, the node discardes the oldest items in the window. Setting this value
+  to zero means that no limit is imposed on the number of items.
+
+The node will retain all events within the time window, provided they do not
+exceed the maximum number.  These config parameters allow the operator to
+loosely regulate how much memory and storage the node allocates to the event
+log.
+
+The event log is shared among all subscribers to the node.
+
+> **Open question:** Should events persist across node restarts?
+>
+> The current event API does not persist events. Some events may be _indexed_,
+> but the node does not keep a complete record: Only BeginBlock and EndBlock
+> events are passed to the indexer, if an indexer is configured at all.
+>
+> The initial implementation will not persist events across restarts.
+
+> **Open question:** The node already stores some event data persistently,
+> notably the events that correspond with block updates. I'm not sure whether
+> it persists other intermediate events, but I think it does not. The event log
+> could be combined with other storage (e.g., reuse the database shared by the
+> blockstore and statestore). Unless we need events to persist, however, it is
+> simpler, easier, and faster to buffer in memory.
+
+### Query API
+
+To retrieve event data, the client will call the (new) RPC method `events`.
+The parameters of this method will correspond to the following Go types:
+
+```go
+type EventParams struct {
+    // Optional filter spec. If nil or empty, all items are eligible.
+    Filter *Filter `json:"filter"`
+
+    // The maximum number of eligible results to return.
+    // If zero or negative, the server will report a default number.
+    MaxResults int `json:"max_results"`
```

Comment:
Which direction does this cut the results in: the earlier events are cut or the latest events?

Reply:
Don't worry I scrolled down and saw you answered it later on

Reply:
I did, but for posterity: One always gets the newest (most recent) matching events.